tortoisebeacon/tortoise_beacon.go

```diff
@@ -606,32 +614,41 @@ func (tb *TortoiseBeacon) runConsensusPhase(ctx context.Context, epoch types.Epo
 	tb.Log.With().Debug("Starting following voting message sender",
 		log.Uint64("epoch_id", uint64(epoch)))
 
+	// we need to pass a map with spacetime unit allowances before any round is started
+	_, atxs, err := tb.atxDB.GetEpochWeight(epoch)
+	if err != nil {
+		tb.Log.With().Panic("can't load list of atxs", log.Err(err))
+	}
+	ua := weakcoin.UnitAllowances{}
+	for _, id := range atxs {
+		header, err := tb.atxDB.GetAtxHeader(id)
+		if err != nil {
+			tb.Log.With().Panic("can't load atx header", log.Err(err))
+		}
+		ua[string(header.NodeID.VRFPublicKey)] += uint64(header.NumUnits)
+	}
+	tb.weakCoin.StartEpoch(epoch, ua)
+
 	// For K rounds: In each round that lasts δ, wait for proposals to come in.
 	// For next rounds,
 	// wait for δ time, and construct a message that points to all messages from previous round received by δ.
 	// rounds 1 to K
 	ticker := time.NewTicker(tb.votingRoundDuration + tb.weakCoinRoundDuration)
 	defer ticker.Stop()
 
-	tb.sendFollowingVotesLoopIteration(ctx, epoch, firstRound+1)
-
-	for round := firstRound + 2; round <= tb.lastPossibleRound(); round++ {
+	for round := firstRound + 1; round <= tb.lastPossibleRound(); round++ {
+		tb.sendFollowingVotesLoopIteration(ctx, epoch, round)
 		select {
 		case <-ticker.C:
-			tb.sendFollowingVotesLoopIteration(ctx, epoch, round)
 		case <-tb.CloseChannel():
 			return
 		case <-ctx.Done():
 			return
 		}
+		tb.weakCoin.CompleteRound()
 	}
 
-	tb.Log.With().Debug("Following voting message sender finished",
-		log.Uint64("epoch_id", uint64(epoch)))
-
-	tb.waitAfterLastRoundStarted()
-	tb.weakCoin.OnRoundFinished(epoch, tb.lastPossibleRound())
-
+	tb.weakCoin.CompleteEpoch()
```

Comment:
@nkryuchkov i didn't understand why do we need to wait twice. in the loop above we are waiting already. was it intentional?

Reply:
@dshulyak The for loop finishes after the last following message is sent, i.e. at the beginning of the last round (sending is done in a goroutine in `sendFollowingVotesLoopIteration`). Then `waitAfterLastRoundStarted` waits for one round until the last round is definitely finished to be sure that all last round votes are received before starting the beacon calculation.

Reply:
Actually, something like `waitUntilLastRoundFinished` would be a much better and less confusing name.

Reply:
sorry right, i changed that. now the loop always sends messages, then waits for voting and weak coin duration, completes round and only then exits. so it wasn't waiting twice before.

Reply:
Yeah, waiting in a loop looks like a more clear solution